internal/providers/github/properties.go

```diff
@@ -0,0 +1,146 @@
+//
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	v1 "github.com/stacklok/minder/pkg/providers/v1"
+
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+type propertyWrapper func(ctx context.Context, ghCli v1.GitHub, name string) (map[string]any, error)
+
+type propertyOrigin struct {
+	keys    []string
+	wrapper propertyWrapper
+}
+
+type propertyFetcher struct {
+	ghCli v1.GitHub
+
+	propertyOrigins []propertyOrigin
+}
+
+var repoPropertyDefinitions = []propertyOrigin{
+	{
+		keys:    []string{"is_private", "is_archived"},
+		wrapper: getRepoWrapper,
+	},
+	{
+		keys:    []string{"properties.languages"},
+		wrapper: getLanguagesWrapper,
+	},
+}
+
+func newRepoPropertyFetcher(ghCli v1.GitHub) *propertyFetcher {
+	return &propertyFetcher{
+		ghCli:           ghCli,
+		propertyOrigins: repoPropertyDefinitions,
+	}
+}
+
+func newPropertyFetcher(ghCli v1.GitHub, entType minderv1.Entity) *propertyFetcher {
+	switch entType {
+	case minderv1.Entity_ENTITY_REPOSITORIES:
+		return newRepoPropertyFetcher(ghCli)
+	}
+
+	return nil
+}
+
+func getRepoWrapper(ctx context.Context, ghCli v1.GitHub, name string) (map[string]any, error) {
+	// TODO: this should be a provider interface, even if private
+	slice := strings.Split(name, "/")
+	if len(slice) != 2 {
+		return nil, errors.New("invalid name")
+	}
+
+	repo, err := ghCli.GetRepository(ctx, slice[0], slice[1])
+	if err != nil {
+		return nil, err
+	}
+
+	repoProps := map[string]any{
+		"is_private":  repo.GetPrivate(),
+		"is_archived": repo.GetArchived(),
+	}
+
+	return repoProps, nil
+}
+
+func getLanguagesWrapper(ctx context.Context, ghCli v1.GitHub, name string) (map[string]any, error) {
+	// TODO: this should be a provider interface, even if private
+	slice := strings.Split(name, "/")
```

Comment:
I think we should have a provider name formatter API

Reply:
I agree. It's fine if it's a TODO for now.

Reply:
this is now tracked in issue #4166